### PR TITLE
Copr: implement sub_datetime_and_duration and sub_datetime_and_string

### PR DESCRIPTION
**Pull request by koushiro**
## What have you changed? (mandatory)

Implement:
- `sub_datetime_and duration`
- `sub_datetime_and_string` 
- `sub_time_datetime_null`

Improve: 
- `add_datetime_and_string`

## What are the type of the changes? (mandatory)

- Improvement

## How has this PR been tested? (mandatory)

unit test.

## Does this PR affect documentation (docs) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -393,6 +390,49 @@ impl ScalarFunc {
...
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let arg0: Cow<'a, Time> = try_opt!(self.children[0].eval_time(ctx, row));
+        let arg1: Cow<'a, [u8]> = try_opt!(self.children[1].eval_string(ctx, row));
+        let s = ::std::str::from_utf8(&arg1)?;
+        let arg1 = MyDuration::parse(&arg1, Time::parse_fsp(s))?;
+        let overflow = Error::overflow("TIME", &format!("({} - {})", &arg0, &arg1));
+        let res = match arg0.into_owned().checked_sub(&arg1) {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Should we consider to convert it to type `Datetime` if it is not? @zz-jason

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -393,6 +390,49 @@ impl ScalarFunc {
...
+        Ok(Some(Cow::Owned(res)))
+    }
+
+    #[inline]
+    pub fn sub_time_datetime_null<'a>(
+        &self,
+        ctx: &mut EvalContext,
+        _row: &[Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        Ok(Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz))))
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: I think we should return None here @zz-jason

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (XuHuaiyu) [reply src/coprocessor/dag/expr/builtin_time.rs]: yes

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (XuHuaiyu) [reply src/coprocessor/dag/expr/builtin_time.rs]: Yes

**Comment by AndreMouche** [comment]
Any update?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: Like that?
```rust
let mut res = ...;
res.set_time_type(TimeType::DateTime)?;
```

**Comment by koushiro** [comment]
@AndreMouche PTAL again.

**Comment by AndreMouche** [comment]
@rleungx @XuHuaiyu PTAL

**Pull request merged**